### PR TITLE
Support repeating table header/footer across pages

### PR DESCRIPTION
**Pull request by unageanu**
- Implemented the specification described in [CSS Tables 3](https://drafts.csswg.org/css-tables-3/#repeated-headers)
- Implementation Details:
  - Add RepetitiveElements class which manages repetitive elements like table headers and footers.
  - Introduced `adapt.layout.BlockFormattingContext` for traditional block layout.

### CODE REVIEW

File: src/vivliostyle/layoututil.js [hunk 1]
```diff
@@ -397,5 +399,183 @@ goog.scope(function() {
...
+            return 1;
+        } else {
+            return 0;
+        }
+    };
+
+    /**
+     * @return {number}
+     */
+    RepetitiveElements.prototype.calculateElementHeight = function() {
```
Comment (kwkbtr) [inline src/vivliostyle/layoututil.js]: Is `calculateElementHeight` an appropriate name?
It seems to me that the value calculated here is some kind of offset, not height of elements.

File: src/adapt/layout.js [hunk 2]
```diff
@@ -204,6 +204,63 @@ adapt.layout.BreakPosition.prototype.findAcceptableBreak = function(column, pena
...
+adapt.layout.AbstractBreakPosition.prototype.retrieveRepetitiveElements = function(nodeContext) {
+    if (!nodeContext) return null;
+    var formattingContext = nodeContext.formattingContext;
+    while (formattingContext) {
+        var repetitiveElements = formattingContext.getRepetitiveElements();
+        if (repetitiveElements) return repetitiveElements;
+        formattingContext = formattingContext.getParent();
+    }
+    return null;
+};
```
Comment (kwkbtr) [inline src/adapt/layout.js:247]: I think we need to take into account multiple `RepetitiveElements` when the fragmented element has multiple ancestors each of which has its own `RepetitiveElements`.
I don't think this issue should be addressed in this PR, but we might need to deal with it later.

File: src/vivliostyle/table.js [hunk 3]
```diff
@@ -1225,19 +1350,43 @@ goog.scope(function() {
...
+        var column = cellFragment.pseudoColumn.column;
+        var cellContentElement = column.element;
+        var cellElement = cellFragment.cellNodeContext.viewNode;
+
+        var cellElementRect = column.clientLayout.getElementClientRect(cellElement);
+        if (vertical) {
+            var width = (cellElementRect.right - column.footnoteEdge - repetitiveElements.calculateElementHeight());
+            adapt.base.setCSSProperty(cellContentElement, "width", width + "px");
+        } else {
+            var height = (column.footnoteEdge - repetitiveElements.calculateElementHeight() - cellElementRect.top);
```
[the remark this answers is not in the document]
Comment (kwkbtr) [reply src/vivliostyle/table.js]: 2, 3 and 4 should have been addressed by acb4f6374e627efe5367ca89aa12fe2ccaf28054.

File: src/vivliostyle/table.js [hunk 4]
```diff
@@ -1225,19 +1350,43 @@ goog.scope(function() {
...
+            adapt.base.setCSSProperty(cellContentElement, "height", height + "px");
```
[the remark this answers is not in the document]
Comment (kwkbtr) [reply src/vivliostyle/table.js]: There is another bug that `vertical-align` for a table cell is not effective when the cell is fragmented (see `table_vertical_align.html` for example).
Using `max-height`/`max-width` solves this bug too.

File: src/vivliostyle/table.js [hunk 5]
```diff
@@ -172,39 +176,65 @@ goog.scope(function() {
...
     InsideTableRowBreakPosition.prototype.findAcceptableBreak = function(column, penalty) {
+        var repetitiveElements =  this.getRepetitiveElements();
+        if (repetitiveElements) repetitiveElements.updateState(penalty);
         if (penalty < this.getMinBreakPenalty())
             return null;
-        var allCellsBreakable = this.getAcceptableCellBreakPositions().every(function(bp) {
-            return !!bp.nodeContext;
+        var cellFragments = this.getCellFragments();
+        var allCellsBreakable = this.getAcceptableCellBreakPositions().every(function(bp, index) {
+            return !!bp.nodeContext && !cellFragments[index].pseudoColumn.isStartNodeContext(bp.nodeContext);
```
[the remark this answers is not in the document]
Comment (unageanu) [reply src/vivliostyle/table.js]: Fixed it in [this commit](https://github.com/vivliostyle/vivliostyle.js/commit/482e9ff526e08771f3c2da127abca9011686d80b).

File: src/vivliostyle/table.js [hunk 6]
```diff
@@ -1225,19 +1350,43 @@ goog.scope(function() {
     };
 
     /**
+     * @override
```
[the remark this answers is not in the document]
Comment (unageanu) [reply src/vivliostyle/table.js]: Fixed it in [this commit](https://github.com/vivliostyle/vivliostyle.js/commit/5624f0d56a088aba59ca3a0f5a94e5fa6b517441).

File: src/vivliostyle/table.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (unageanu) [reply src/vivliostyle/table.js]: Fixed it in [this commit](https://github.com/vivliostyle/vivliostyle.js/commit/a8ce0644eb4e50d2eac0fdc321ec28d1c10591fc).

File: src/vivliostyle/layoututil.js (hunk 1 above)
Comment (unageanu) [reply src/vivliostyle/layoututil.js]: Fixed it in [this commit](https://github.com/vivliostyle/vivliostyle.js/commit/b67a229a170f40dbd286033eb8062f2eb4eb64ac).

File: src/vivliostyle/table.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (unageanu) [reply src/vivliostyle/table.js]: Fixed it in [this commit](https://github.com/vivliostyle/vivliostyle.js/pull/319/commits/72dafa12396464670640530f30965d28c026d235).

**Pull request merged**